### PR TITLE
Implementation of gerrit trigger plugin to JJW

### PR DESCRIPTION
**Pull request by erenatas**
Signed-off-by: Eren ATAS <eatas.contractor@libertyglobal.com>

### CODE REVIEW

File: jenkins_job_wrecker/modules/triggers.py [hunk 1]
```diff
@@ -1,4 +1,6 @@
 # encoding=utf8
+from builtins import NotImplementedError
```
Comment (romanek-adam) [inline jenkins_job_wrecker/modules/triggers.py]: builtins is not available in Python 2.7 (tox confirms this)

in any case you don't need it - it works without importing NotImplementedError explicitly

File: jenkins_job_wrecker/modules/triggers.py [hunk 2]
```diff
@@ -38,18 +40,211 @@ def reversebuildtrigger(top, parent):
...
+            raise NotImplementedError("Not implemented file path type: ", file_path_type.tag)
+    return file_paths
+
+
 def gerrittrigger(top, parent):
-    pass
+    gerrit_trigger = {}
+    for child in top:
+        print(child.tag)
+        if child.tag == "gerritProjects":
```
Comment (romanek-adam) [inline jenkins_job_wrecker/modules/triggers.py]: I'd extract this part into a separate function for readability, also the same for "triggerOnEvents"

File: jenkins_job_wrecker/modules/triggers.py [hunk 3]
```diff
@@ -38,18 +40,211 @@ def reversebuildtrigger(top, parent):
...
+                if attribute.text == "onSuccessful":
+                    skip_vote["successful"] = get_bool(attribute.text)
+                if attribute.text == "onFailed":
+                    skip_vote["failed"] = get_bool(attribute.text)
+                if attribute.text == "onUnstable":
+                    skip_vote["unstable"] = get_bool(attribute.text)
+                if attribute.text == "onNotBuilt":
+                    skip_vote["notbuilt"] = get_bool(attribute.text)
+            gerrit_trigger["skip_vote"] = skip_vote
+        elif child.tag == "gerritBuildStartedVerifiedValue":
```
Comment (romanek-adam) [inline jenkins_job_wrecker/modules/triggers.py]: I'd reduce the repetitive work here like in my commit: https://github.com/romanek-adam/jenkins-job-wrecker/commit/2b527399d6866439b0d4e0acc3d247b4b05f4875

File: jenkins_job_wrecker/modules/triggers.py [hunk 4]
```diff
@@ -38,18 +40,211 @@ def reversebuildtrigger(top, parent):
...
+                    trigger_on.append("wip-state-changed-event")
+            gerrit_trigger["trigger-on"] = trigger_on
+        elif child.tag == "dynamicTriggerConfiguration":
+            gerrit_trigger["dynamic-trigger-enabled"] = get_bool(child.text)
+        elif child.tag == "triggerConfigURL":
+            gerrit_trigger["dynamic-trigger-url"] = child.text
+        elif child.tag == "gerritTriggerTimerTask":
+            pass
+        elif child.tag == "triggerInformationAction":
+            pass
```
Comment (romanek-adam) [inline jenkins_job_wrecker/modules/triggers.py]: please document what's the plan for this and other elements that are skipped

File: jenkins_job_wrecker/modules/triggers.py [hunk 5]
```diff
@@ -38,18 +40,211 @@ def reversebuildtrigger(top, parent):
...
-            pass    # TODO
+            pass  # TODO
         else:
             raise NotImplementedError('cannot handle reverse trigger '
                                       'setting %s' % child.tag)
 
     parent.append({'reverse': reverse})
 
 
+def process_file_paths(attribute):
```
Comment (romanek-adam) [inline jenkins_job_wrecker/modules/triggers.py]: I'd prefix the function with 'gerrit_'

File: tests/test_modules_triggers.py [hunk 6]
```diff
@@ -0,0 +1,20 @@
+# -*- coding: utf-8 -*-
+from jenkins_job_wrecker.cli import get_xml_root, root_to_yaml
+from jenkins_job_wrecker.modules.triggers import gerrittrigger
+import os, yaml
+
+fixtures_path = os.path.join(os.path.dirname(__file__), 'fixtures', 'triggers')
+
+
+class TestGerritTrigger(object):
+    def test_basic(self):
```
Comment (erenatas) [inline tests/test_modules_triggers.py]: The name of the test func should be changed

File: jenkins_job_wrecker/modules/triggers.py [hunk 7]
```diff
@@ -38,18 +40,211 @@ def reversebuildtrigger(top, parent):
...
+                    trigger_on.append({"comment-added-event": comment_added_event})
+                elif event.tag == sonyericsson_prefix + "PluginCommentAddedContainsEvent":
+                    trigger_on.append({"comment-added-contains-event": {"comment-contains-value": event[0].text}})
+                elif event.tag == sonyericsson_prefix + "PluginDraftPublishedEvent":
+                    trigger_on.append("draft-published-event")
+                elif event.tag == sonyericsson_prefix + "PluginPatchsetCreatedEvent":
+                    patchset_created_event = {}
+                    for attribute in event:
+                        if attribute.tag == "excludeDrafts":
+                            patchset_created_event["exclude-draft"] = get_bool(attribute.text)
```
Comment (erenatas) [inline jenkins_job_wrecker/modules/triggers.py]: It should be "exclude-drafts"

File: tests/fixtures/triggers/gerrit_trigger.yml [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+            approval-category: CRVW
+            approval-value: zxcz
+        - comment-added-event:
+            approval-category: VRIF
+            approval-value: aasdasd
+        - comment-added-contains-event:
+            comment-contains-value: reggex
+        - draft-published-event
+        - patchset-created-event:
+            exclude-draft: true
```
Comment (erenatas) [inline tests/fixtures/triggers/gerrit_trigger.yml]: Also should be "exclude-drafts" here.

File: jenkins_job_wrecker/modules/triggers.py (hunk 5 above)
Comment (erenatas) [reply jenkins_job_wrecker/modules/triggers.py]: I will also make the funcs private with double underscore prefix.

**Comment by romanek-adam** [comment]
I think you should squash the commits into one.

**Comment by erenatas** [comment]
Updated.

**Review by vdanaila — APPROVED** [review]
Looks good, tested locally.

File: jenkins_job_wrecker/modules/triggers.py [hunk 9]
```diff
@@ -38,18 +39,192 @@ def reversebuildtrigger(top, parent):
...
+            for attribute in child:
+                if attribute.text == "onSuccessful":
+                    skip_vote["successful"] = get_bool(attribute.text)
+                if attribute.text == "onFailed":
+                    skip_vote["failed"] = get_bool(attribute.text)
+                if attribute.text == "onUnstable":
+                    skip_vote["unstable"] = get_bool(attribute.text)
+                if attribute.text == "onNotBuilt":
+                    skip_vote["notbuilt"] = get_bool(attribute.text)
+            gerrit_trigger["skip_vote"] = skip_vote
```
Comment (romanek-adam) [inline jenkins_job_wrecker/modules/triggers.py]: Shouldn't `skip_vote` be `skip-vote` (the key)? I'm wondering how did the tests pass this? The test XML probably contains the default values...

**Review by romanek-adam — CHANGES_REQUESTED** [review]

File: jenkins_job_wrecker/modules/triggers.py (hunk 9 above)
Comment (erenatas) [reply jenkins_job_wrecker/modules/triggers.py]: That shouldn't have happened :/ I will handle it.

Comment (romanek-adam) [reply jenkins_job_wrecker/modules/triggers.py]: I think we should have a test which fails for the old implementation which uses "skip_vote" key, right?

**Review by romanek-adam — APPROVED** [review]

**Pull request merged**